### PR TITLE
fix(selectors): use Maps instead of objects

### LINKED ISSUE #12351: Can't use css class "constructor"
If I have css class "constructor" defined in my template alike
```
<div class="constructor"></div>
```
I got an error
```
Unhandled Promise rejection: Cannot read property 'finalize' of undefined ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read property 'finalize' of undefined(…) TypeError: Cannot read property 'finalize' of undefined
    at SelectorMatcher._matchTerminal
    at SelectorMatcher.match
```

It's because of that line
https://github.com/angular/angular/blob/master/modules/%40angular/compiler/src/selector.ts#L300

Empty map object still IS an object, hence it has constructor and it's successfully matched via selector `map["constructor"]`

To fix this issue I need to rename css class to other name.

Tested with angular 2.0.2 and from source it seems still actual in 2.1.0

-
I'm too lazy to fill up your forms and etc, you might close my issue without investigating.
It's just a notice of unobvious side-effect that takes some time with debugging ng2 code.

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/12351

### CODE REVIEW

**Comment by vicb** [comment]
@DzmitryShylovich Thanks for your PR !!

Would you have time to benchmark:
- `Maps` vs `{}` (read and write, native & polyfill),
- creating a `Map` vs `Object.create(null)`.

That would be really great !

**Comment by vicb** [comment]
(restarted the CI, probably only a flake)

File: modules/@angular/compiler/src/selector.ts [hunk 1]
```diff
@@ -135,12 +135,14 @@ export class SelectorMatcher {
     return notMatcher;
   }
 
-  private _elementMap: {[k: string]: SelectorContext[]} = {};
-  private _elementPartialMap: {[k: string]: SelectorMatcher} = {};
-  private _classMap: {[k: string]: SelectorContext[]} = {};
-  private _classPartialMap: {[k: string]: SelectorMatcher} = {};
-  private _attrValueMap: {[k: string]: {[k: string]: SelectorContext[]}} = {};
-  private _attrValuePartialMap: {[k: string]: {[k: string]: SelectorMatcher}} = {};
+  private _elementMap: Map<string, SelectorContext[]> = new Map<string, SelectorContext[]>();
```
Comment (vicb) [inline modules/@angular/compiler/src/selector.ts]: you don't need to type both sides.

**Comment by vicb** [comment]
could you please fix the nits and rebase

**Comment by DzmitryShylovich** [comment]
> Would you have time to benchmark:

jsperf is down :(

**Comment by vicb** [comment]
Thanks !

**Comment by vsavkin** [comment]
Merged https://github.com/angular/angular/commit/d321b0ebf5fa020ac7aecd4d718479a862e0b920

**Pull request closed without merging**